git: worktree, Optimize Checkout for same tree

cedric-appdirect opened the pull request:
Add fast-path optimization to Checkout() that skips expensive worktree scanning when target commit has the same tree as current HEAD.

This matches git CLI behavior which only updates files when trees differ. When creating a new branch from current HEAD (same tree), this eliminates unnecessary file enumeration and hashing that can take seconds on large repositories.

Implementation:
- Add canSkipWorktreeUpdate() helper that compares tree hashes
- Fast path applies when: !Force && !Keep && no sparse directories
- Falls back to existing Reset() logic for different trees
- Graceful error handling with fallback to slow path

Testing:
- Added unit tests for fast path scenarios
- Added benchmarks demonstrating improvement
- All existing checkout tests pass
- No breaking changes to API

cedric-appdirect committed:
git: worktree, Optimize Checkout for same tree

Add fast-path optimization to Checkout() that skips expensive worktree
scanning when target commit has the same tree as current HEAD.

This matches git CLI behavior which only updates files when trees differ.
When creating a new branch from current HEAD (same tree), this eliminates
unnecessary file enumeration and hashing that can take seconds on large
repositories.

Implementation:
- Add canSkipWorktreeUpdate() helper that compares tree hashes
- Fast path applies when: !Force && !Keep && no sparse directories
- Falls back to existing Reset() logic for different trees
- Graceful error handling with fallback to slow path

Testing:
- Added unit tests for fast path scenarios
- Added benchmarks demonstrating improvement
- All existing checkout tests pass
- No breaking changes to API

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, ensure that the worktree match the target tree.

This will verify that their is actually file in the worktree and that it
is properly populated. This would have triggered an issue for tests that
use memfs.New() or after NoCheckout clone.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, we should pass the resolved commit as the head.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

pjbgf commented on worktree_bench_test.go:
```diff
@@ -0,0 +1,150 @@
+
+	// Benchmark creating new branches from current HEAD (all same tree)
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
```
```suggestion
	for b.Loop() {
```

pjbgf commented on worktree_bench_test.go:
```diff
@@ -0,0 +1,150 @@
+// BenchmarkCheckoutSameTree benchmarks the common case of creating a new
+// branch from current HEAD, which has the same tree.
+// This is the primary use case that benefits from fast path optimization.
+func BenchmarkCheckoutSameTree(b *testing.B) {
```
There is quite a bit of code duplication around the benchmarks, please sub-benchmarks to decrease the code verbosity. [Here](https://github.com/go-git/go-git/blob/d11db7d85708cb0fb40abba8973ef4937fa4342a/plumbing/objectid_test.go#L200)'s an example.

cedric-appdirect replied:
Oh, didn't know about that way of writing benchmark. Thanks, will do.

pjbgf reviewed (changes requested):
@cedric-appdirect thanks for looking into this. Please make the changes on the benchmarks. For the other tests we could do on a follow-up PR.

Please add the results of your benchmarks into the PR body for future reference.

ferhatelmas commented on worktree.go:
```diff
@@ -286,6 +307,51 @@ func (w *Worktree) setHEADToBranch(branch plumbing.ReferenceName, commit plumbin
+// canSkipWorktreeUpdate determines if we can skip the worktree update
+// during checkout by comparing tree hashes. Returns true if current HEAD
+// and target commit point to the same tree, meaning no files need to change.
+func (w *Worktree) canSkipWorktreeUpdate(targetCommit plumbing.Hash) (bool, error) {
```
```suggestion
func (w *Worktree) canSkipWorktreeUpdate(targetCommitHash plumbing.Hash) (bool, error) {
```

to be consistent with `indexMatchesTargetTree`

cedric-appdirect committed:
git: worktree, refactor checkout benchmark suggested by @pjbgf

Use b.Loop and b.Run to reduce code duplication and ensure correctness.

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>
Signed-off-by: Cedric Bail <133815608+cedric-appdirect@users.noreply.github.com>
Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, rename parameter to be consistent with indexMatchesTargetTree

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, ensure Checkout fast pass lookup CommitObject once.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect commented on worktree.go:
```diff
@@ -183,6 +183,27 @@ func (w *Worktree) Checkout(opts *CheckoutOptions) error {
+		if canUseFastPath, err := w.canSkipWorktreeUpdate(c); err == nil && canUseFastPath {
+			if indexMatches, err := w.indexMatchesTargetTree(c); err == nil && indexMatches {
+				if !opts.Hash.IsZero() && !opts.Create {
+					return w.setHEADToCommit(c)
```
Yes, I had to switch to use the "resolved" hash otherwise worktree_test.go fail at line 960 with error "object not found".

cedric-appdirect committed:
git: worktree, ensure TestCheckoutFastPathVerification trigger the checkout fast path

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, update documentation pointing to git cli code.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect commented on worktree.go:
```diff
@@ -286,6 +307,51 @@ func (w *Worktree) setHEADToBranch(branch plumbing.ReferenceName, commit plumbin
+
+// indexMatchesTargetTree checks if the current index already matches the target tree.
+// This is the proper check to determine if we can skip worktree updates during checkout.
+// Git uses the index (staging area) as the source of truth - if index == target tree,
```
Their is no documentation about this, just code. I have added a link to the code. There is also a bunch of related commit which span a few years. Not to sure what is the best choice here.

pjbgf commented on worktree_bench_test.go:
```diff
@@ -0,0 +1,132 @@
+func benchmarkCheckoutSameTree(b *testing.B, w *Worktree) {
+	i := 0
+	for b.Loop() {
+		branchName := plumbing.NewBranchReferenceName(fmt.Sprintf("bench-branch-%d", i))
```
```suggestion
		branchName := plumbing.NewBranchReferenceName(fmt.Sprintf("bench-branch-%d", i++))
```

cedric-appdirect committed:
git: worktree, fail benchmark instead of skip in case of setup error

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, run benchmark without build tag restriction

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, simplify benchmark function prototype.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, remove unecessary higher level tests for fast path

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, add unit tests for canSkipWorktreeUpdate and indexMatchesTargetTree

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, address benchmark lint issue and modernization

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, checkout, Fix CI and clarify fast-path logic for same-tree checkout.

Update go-git-fixtures import from v5 to v6 to match upstream migration, handle v6
API changes (DotGit() now returns error), and fix benchmark assertions. Expand comment
documentation explaining why ErrUnstagedChanges is intentionally not checked when trees
match (dirty worktree files carry over untouched, same as git CLI), why opts.Create is
safe with the fast path, and clarify git reference.

Assisted-by: OpenCode with Claude Opus 4.6 <noreply@anthropic.com>
Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>
